In [1]:
from tqdm import tqdm
import pandas as pd
import re

In [2]:
df_sent_comp=pd.read_excel('../output/qsw_subtract.xlsx')

### Male kinship extraction.

In [3]:
re_rule_list=[r'(三十一世祖|十八世祖|十六代祖|十五代祖|十三代祖|十二世祖|十二代祖|十一代祖|十一世祖|四代祖母|曾祖王父|外曾大父|外曾王父|外曾祖父|六代祖|十代祖|十世祖|九世祖|八世祖|七世祖|六世祖|五世祖|五代祖|四世祖|四代祖|高祖母|曾祖考|曾大父|大王父|曾王父|皇曾祖|曾大考|曾伯祖|曾叔祖|先祖考|先大父|先王父|皇祖考|伯祖父|叔祖父|從伯祖|再從伯|再從叔|再從舅|從伯父|從叔父|外祖父|外王父|外大父|外叔祖|外伯祖|再從舅|外伯父|外叔父|外大祖|外曾祖|遠祖|其先|高祖|曾祖|先祖|大父|王父|大考|從祖|族祖|伯祖|叔祖|皇考|先父|先考|先君|伯父|世父|叔父|族父|從伯|從父|從叔|從舅|外舅|母舅|外祖|從舅|外舅|祖|父|考|舅|舅)[諱|曰](.{0,2}?)/',
              r'^(三十一世祖|十八世祖|十六代祖|十五代祖|十三代祖|十二世祖|十二代祖|十一代祖|十一世祖|四代祖母|曾祖王父|外曾大父|外曾王父|外曾祖父|六代祖|十代祖|十世祖|九世祖|八世祖|七世祖|六世祖|五世祖|五代祖|四世祖|四代祖|高祖母|曾祖考|曾大父|大王父|曾王父|皇曾祖|曾大考|曾伯祖|曾叔祖|先祖考|先大父|先王父|皇祖考|伯祖父|叔祖父|從伯祖|再從伯|再從叔|再從舅|從伯父|從叔父|外祖父|外王父|外大父|外叔祖|外伯祖|再從舅|外伯父|外叔父|外大祖|外曾祖|遠祖|高祖|曾祖|先祖|大父|王父|大考|從祖|族祖|伯祖|叔祖|皇考|先父|先考|伯父|世父|叔父|族父|從伯|從父|從叔|從舅|外舅|母舅|外祖|從舅|外舅|祖|父|考|舅)([^居某早丧卒母妣諱曰也]{0,2})/',
              r'[公君吾夫人我/](三十一世祖|十八世祖|十六代祖|十五代祖|十三代祖|十二世祖|十二代祖|十一代祖|十一世祖|四代祖母|曾祖王父|外曾大父|外曾王父|外曾祖父|六代祖|十代祖|十世祖|九世祖|八世祖|七世祖|六世祖|五世祖|五代祖|四世祖|四代祖|高祖母|曾祖考|曾大父|大王父|曾王父|皇曾祖|曾大考|曾伯祖|曾叔祖|先祖考|先大父|先王父|皇祖考|伯祖父|叔祖父|從伯祖|再從伯|再從叔|再從舅|從伯父|從叔父|外祖父|外王父|外大父|外叔祖|外伯祖|再從舅|外伯父|外叔父|外大祖|外曾祖|遠祖|高祖|曾祖|先祖|大父|王父|大考|從祖|族祖|伯祖|叔祖|皇考|先父|先考|伯父|世父|叔父|族父|從伯|從父|從叔|從舅|外舅|母舅|外祖|從舅|外舅|祖|父|考|舅)([^居某早丧卒母妣諱曰也]{0,2}?)/',
              r'[公君吾夫人我/](三十一世祖|十八世祖|十六代祖|十五代祖|十三代祖|十二世祖|十二代祖|十一代祖|十一世祖|四代祖母|曾祖王父|外曾大父|外曾王父|外曾祖父|六代祖|十代祖|十世祖|九世祖|八世祖|七世祖|六世祖|五世祖|五代祖|四世祖|四代祖|高祖母|曾祖考|曾大父|大王父|曾王父|皇曾祖|曾大考|曾伯祖|曾叔祖|先祖考|先大父|先王父|皇祖考|伯祖父|叔祖父|從伯祖|再從伯|再從叔|再從舅|從伯父|從叔父|外祖父|外王父|外大父|外叔祖|外伯祖|再從舅|外伯父|外叔父|外大祖|外曾祖|遠祖|高祖|曾祖|先祖|大父|王父|大考|從祖|族祖|伯祖|叔祖|皇考|先父|先考|伯父|世父|叔父|族父|從伯|從父|從叔|從舅|外舅|母舅|外祖|從舅|外舅|祖|父|考|舅)([^居某早丧卒母妣諱曰也]{0,2}?)$',
              r'(三十一世祖|十八世祖|十六代祖|十五代祖|十三代祖|十二世祖|十二代祖|十一代祖|十一世祖|四代祖母|曾祖王父|外曾大父|外曾王父|外曾祖父|六代祖|十代祖|十世祖|九世祖|八世祖|七世祖|六世祖|五世祖|五代祖|四世祖|四代祖|高祖母|曾祖考|曾大父|大王父|曾王父|皇曾祖|曾大考|曾伯祖|曾叔祖|先祖考|先大父|先王父|皇祖考|伯祖父|叔祖父|從伯祖|再從伯|再從叔|再從舅|從伯父|從叔父|外祖父|外王父|外大父|外叔祖|外伯祖|再從舅|外伯父|外叔父|外大祖|外曾祖|遠祖|其先|高祖|曾祖|先祖|大父|王父|大考|從祖|族祖|伯祖|叔祖|皇考|先父|先考|先君|伯父|世父|叔父|族父|從伯|從父|從叔|從舅|外舅|母舅|外祖|從舅|外舅|祖|父|考|舅)(某官{0,1}?)',
              r'(三十一世祖|十八世祖|十六代祖|十五代祖|十三代祖|十二世祖|十二代祖|十一代祖|十一世祖|四代祖母|曾祖王父|外曾大父|外曾王父|外曾祖父|六代祖|十代祖|十世祖|九世祖|八世祖|七世祖|六世祖|五世祖|五代祖|四世祖|四代祖|高祖母|曾祖考|曾大父|大王父|曾王父|皇曾祖|曾大考|曾伯祖|曾叔祖|先祖考|先大父|先王父|皇祖考|伯祖父|叔祖父|從伯祖|再從伯|再從叔|再從舅|從伯父|從叔父|外祖父|外王父|外大父|外叔祖|外伯祖|再從舅|外伯父|外叔父|外大祖|外曾祖|遠祖|其先|高祖|曾祖|先祖|大父|王父|大考|從祖|族祖|伯祖|叔祖|皇考|先父|先考|先君|伯父|世父|叔父|族父|從伯|從父|從叔|從舅|外舅|母舅|外祖|從舅|外舅|祖|父|考|舅)[諱|曰](.{0,2})$',
             ]

In [4]:
df_sent_comp.sample(3)

content_id subject author  \
50728     5314621     陳知和    晁補之   
58344     5174372     趙克凝     沈括   
41802     5517068      曾耑     胡銓   

                                                    sent  \
50728  女長適朝散郎潘行/wsep/次適通直郎王鎮/wsep/次適道州録事楚沖/wsep/次適宣德郎...   
58344  臣某謹按/wm/南康侯/wsep/故虔州觀察使/wsep/南康侯承睦之子/wsep/廣陵郡王...   
41802  先是從父文林君之子卒/wsep/無子/wsep/二女在室/wsep/君經紀喪事/wsep/雖...   

                                                 content  \
50728  君陳氏，諱知和，字德時，按其譜，博州人也。六世祖翔，唐末從事成都府。王建有異志，翔屢沮之，廢...   
58344  公諱克凝，字希政。熙寧八年閏四月，公以疾聞，甲午卒。詔以虔州觀察使、南康侯告於第。冬十月己酉...   
41802  君諱耑，字景初，姓曾氏，吉之廬陵人。曾王父諱億，贈朝議大夫。王父諱鎮，左朝散大夫、廣南西路轉...   

                                               sent_comp   nz            ns  \
50728  女長適/no_noc/潘行/wsep/次適/no_noc/王鎮/wsep/次適/ns//no...  NaN            道州   
58344  臣某謹按/wm//ns/侯/wsep//vno//ns//no_noc//wsep//ns/...  NaN  廣陵郡#南康#南康#虔州   
41802  先是從父文林君之子卒/wsep/無子/wsep/二女在室/wsep/君經紀喪事/wsep/雖...  NaN           NaN   

               no_noc  vno  
50728  朝散郎#宣德郎#通直郎#録事  NaN  
58344             觀察使    故  
41802             NaN  NaN

In [5]:
kin_list=[]
df_kin_extracted=pd.DataFrame()
for re_rule in re_rule_list:
    for index in tqdm(df_sent_comp.index):
        string=df_sent_comp.loc[index, 'sent_comp']
        content_id=df_sent_comp.loc[index, 'content_id']
        content=df_sent_comp.loc[index, 'content']
        author=df_sent_comp.loc[index, 'author']
        subj=df_sent_comp.loc[index, 'subject']
        kin_list_temp=re.findall(re_rule, string)
        if kin_list_temp!=[]:
            df_kin_extracted=pd.concat([pd.DataFrame([[index, content_id, content, 
                                                       author, subj, kin[0], kin[1]] 
                                                      for kin in kin_list_temp]), df_kin_extracted], ignore_index=True)

100%|██████████| 57733/57733 [00:06<00:00, 8432.30it/s]


In [8]:
df_kin_extracted.columns=['sent_comp_index', 'content_id', 'content', 'author', 'subject', 'kin', 'name']
df_kin_extracted.sample(3)

sent_comp_index  content_id  \
5764            10471     5465520   
7977            22854     5870209   
8402             2322     5870340   

                                                content author subject kin  \
5764  　　曾祖伯照，故贈光禄卿。祖母：扶風太君魯氏，彭城太君錢氏。祖迪，故任尚書都官郎中致仕。祖母...     程瑀     NaN   父   
7977  吾鄉城南楊氏最盛。先正獻公絜齋先生久處其家塾，子弟多秀士，服習詩書，頡頏場屋，聲譽大振。子美...     袁甫      何璘   祖   
8402  君諱夔友，字子韶，姓吕氏，世爲齊之歷城人。曾祖諱頤浩，中興名宰相，贈太師秦國公，配饗高宗廟庭...     袁甫     吕夔友   祖   

     name  
5764   天民  
7977    萃  
8402    摭

In [9]:
for index in tqdm(df_kin_extracted.index):
    sent_comp_index = df_kin_extracted.loc[index, 'sent_comp_index']
    df_temp=df_sent_comp.loc[sent_comp_index][['sent', 'sent_comp', 'nz','ns','no_noc','vno']]
    df_kin_extracted.loc[index, 'sent']=df_temp['sent']
    df_kin_extracted.loc[index, 'sent_comp']=df_temp['sent_comp']
    df_kin_extracted.loc[index, 'nz']=df_temp['nz']
    df_kin_extracted.loc[index, 'ns']=df_temp['ns']
    df_kin_extracted.loc[index, 'no_noc']=df_temp['no_noc']
    df_kin_extracted.loc[index, 'vno']=df_temp['vno']

100%|██████████| 8477/8477 [01:04<00:00, 130.99it/s]


In [10]:
df_kin_extracted.sample(3)

sent_comp_index  content_id  \
2635            11526     5307283   
1276            62232     5131852   
6516            67341     5062315   

                                                content author subject kin  \
2635  寶元、康定間，屬羌叛亂，西邊苦兵，天子恤勞吏民，制詔二府，謀于衆庶，以協大同。士争論戰守，計...    陳師道     曾易占   祖   
1276  公諱温瑜，字子美，御史中丞兼宗正卿、贈太師、中書令兼尚書令魏文定公諱安仁之長子。魏公在真宗世...     蘇頌     趙温瑜  曾祖   
6516  君諱寶臣，字元珍，姓丁氏，常州晋陵人也。景祐元年，舉進士及第，爲峽州軍事判官，淮南節度掌書記...    歐陽修     丁寶臣   祖   

     name                                               sent  \
2635   延鐸                                     公祖延鐸/wsep/散騎常侍   
1276       公世爲河南洛陽人/wsep/曾祖贈太尉諱處厚/wsep/祖殿中侍御史/wsep/贈太師/ws...   
6516    某                             曾祖諱某/wsep/祖諱某/wsep/皆不仕   

                                              sent_comp   nz        ns  \
2635                                 公祖延鐸/wsep//no_noc/  NaN       NaN   
1276  公世/vno//ns//ns/人/wsep/曾祖/vno//no_noc/諱處厚/wsep/...  NaN  洛陽#洛陽#河南   
6516                             曾祖諱某/wsep/祖諱某/wsep/皆不仕  NaN       NaN   

               no_noc    vno  
2635             散騎常侍    NaN  
1276  殿中侍御史#中書令#太師#太尉  贈#贈#爲  
6516              NaN    NaN

In [11]:
df_kin_extracted[df_kin_extracted['name']!=''].sample(5)

sent_comp_index  content_id  \
2447            19549     5934052   
3472            55868     5234554   
2659            10573     5453312   
330             53184     5264642   
1300            61024     5141203   

                                                content author    subject kin  \
2447  國家設科發策，正以伸天下敢言之氣，一有拘忌，有司反先喑無聲。嘻，可歎已!然於斯時也，有能獨諤...     黄震        陸鵬升   父   
3472  君劉氏，開封人。曾祖懷懿，南作坊副使。祖用可，閤門祗候，贈左千牛衛將軍。父漢卿，內殿崇班。母...    范祖禹  劉漢卿女、趙令振妻   父   
2659  資政殿學士、權同知三省樞密院事劉公自贛上走使數千里，狀其先宫師之行，寓書于某曰：‘珏不天，逮...     張守        劉定國   祖   
330   嗚呼，世人嘗知按學文以求君子，而不知君子之盛，有不由於學問者，其於所求可謂與余異矣。盛古之時...    吕南公        NaN   祖   
1300  真州司法京兆杜涣濟叔，年三十七，以皇祐四年四月辛酉卒。子男某尚稚，自將以下合貨財以葬於北城之...    王安石         杜涣   父   

     name                                              sent  \
2447   守謙         曾祖憲遷蕪湖/wsep/祖宗周遷旴江/wsep/父守謙贈宣教郎/wsep/始遷臨川   
3472   漢卿                                     父漢卿/wsep/內殿崇班   
2659   承福  曾祖文奎不仕/wsep/祖承福贈太子少保/wsep/父涉贈太子少傅/wsep/母陳氏贈齊安郡夫人   
330     某                     曾祖某/wsep/祖某/wsep/父某/wsep/皆已諱名   
1300   詢嘗        杜氏世占永寧之博野/wsep/父詢嘗歷江寧府司録參軍/wsep/遂葬/wsep/家焉   

                                              sent_comp   nz         ns  \
2447  曾祖憲/vno//ns//wsep/祖宗周遷旴江/wsep/父守謙/vno//no_noc/...  NaN      蕪湖#臨川   
3472                                  父漢卿/wsep//no_noc/  NaN        NaN   
2659  曾祖文奎不仕/wsep/祖承福/vno//no_noc//wsep/父涉/vno//no_n...  NaN        齊安郡   
330                       曾祖某/wsep/祖某/wsep/父某/wsep/皆已諱名  NaN        NaN   
1300  杜氏世占/ns/之/ns//wsep/父詢嘗/vno//ns/司録/no_noc//wsep...  NaN  江寧府#永寧#博野   

         no_noc    vno  
2447        宣教郎  贈#遷#遷  
3472       內殿崇班    NaN  
2659  太子少保#太子少傅  贈#贈#贈  
330         NaN    NaN  
1300         參軍      歷

In [12]:
df_kin_extracted[df_kin_extracted['name']!=''].drop('sent_comp_index', axis=1).sort_values(by='content_id').to_excel('../output/df_qsw_kin_extracted_male.xlsx', encoding='utf8')

### Female kinship extraction.

In [33]:
re_rule_list=[r'(外叔祖母|外伯祖母|曾祖母|曾祖妣|曾大母|曾王母|曾祖姑|大王母|從祖母|繼祖母|叔祖母|從祖姑|伯祖母|叔祖母|再從姑|從伯姑|外王母|外祖母|祖母|王母|大母|祖妣|祖姑|伯母|叔母|仲姑|從姑|皇妣|先妣|後母|繼母|嫡母|生母|庶母|族母|姨母|外姑|從妗|外姑|姑|母|妣|姨|姑)([^曰也].{0,20}?)氏',
              r'([始先早前首再次又後繼]?娶|取|配)(.{1,10}?)氏',
              r'(繼)([^母].{0,10}?)氏',
              r'(妻)(.{0,10}?)氏',
             ]

In [34]:
kin_list=[]
df_kin_extracted=pd.DataFrame()
for re_rule in re_rule_list:
    for index in tqdm(df_sent_comp.index):
        string=df_sent_comp.loc[index, 'sent_comp']
        content_id=df_sent_comp.loc[index, 'content_id']
        content=df_sent_comp.loc[index, 'content']
        author=df_sent_comp.loc[index, 'author']
        subj=df_sent_comp.loc[index, 'subject']
        kin_list_temp=re.findall(re_rule, string)
        if kin_list_temp!=[]:
            for kin_nm in kin_list_temp:
                if kin_nm[1]!='':
                    kin_nm=(kin_nm[0], kin_nm[1][-1])
                df_kin_extracted=pd.concat([pd.DataFrame([[index, content_id, content, author, subj, re_rule, kin_nm[0], kin_nm[1]]]), df_kin_extracted], ignore_index=True)
#                 else:
#                     print(string, kin_list_temp)

100%|██████████| 57733/57733 [00:06<00:00, 8577.89it/s]


In [35]:
df_kin_extracted.columns=['sent_comp_index', 'content_id', 'content', 'author', 'subject', 're_rule', 'kin', 'name']

In [36]:
df_kin_extracted.sample(10)

sent_comp_index  content_id  \
3553            64244     5111796   
1687            41993     5516988   
2166            31499     5749322   
4295            40576     5576105   
1993            35169     5687065   
2678            18302     8707177   
5831              198     5578423   
73              50408     5314745   
2708            17499     8712147   
4846            23727     5840704   

                                                content author subject  \
3553  公諱亢，字興宗，丹陽人。幼聰發過人，方十歲時，已能日誦書五千言。其爲賦詩縱豪，自鄉里先生見皆...     王珪     NaN   
1687  某自海南内徙合江，家弟鎬渠陽判官罷歸，道湘中，留合江踰累朔。舂容及里中老，僂數至蘭溪曾積臣，...     胡銓     曾敏遜   
2166  思正，姓陳氏，諱端中，思正其字也，世爲婺之永康人。曾祖博，祖回，父子茂，皆不顯於世。余嘗銘陳...     陳亮     陳端中   
4295  龍圖閣學士、新安郡侯、贈左通奉大夫羅公諱汝楫之夫人，湖州安吉俞氏女也。曾祖諱汝尚，屯田郎中，...     洪适    羅汝楫妻   
1993  公姓劉氏，諱子羽，字彦修。世爲京兆人，八世祖避五季之亂，徙家建州。曾祖太素，贈朝議大夫；祖民...     張栻     劉子羽   
2678  〔蓋文〕大宋故光禄寺丞雷公墓誌銘〔誌文〕大宋故光禄寺丞雷公墓誌銘并序兄將仕郎守秘書省著作佐郎...    雷孝先     雷孝孫   
5831  宋故知永康軍柳公佾，字公雅，家世秦州成紀縣。曾祖英，贈千牛衛將軍；妣徐氏、劉氏，皆令人。祖淵...    王汝霖     NaN   
73    進士李君公裕，字好問，濟州金鄉人。尚書工部郎中、直史館定之孫，而奉議郎、通判潞州事衆之子也。...    晁補之     NaN   
2708  元祐二年夏四月，忱適友人之居，日夕而歸。府君之孫玘踵□□者三詰朝，左藏君至，已而玘又至。忱曰...     高忱     NaN   
4846  始予在三山，聞同年李肖望、友人黄仲玉皆號爲長者家。仲玉居城東，其五世祖以寬垕樂施予得是名閭里...    真德秀     李起渭   

                                                re_rule kin name  
3553  (外叔祖母|外伯祖母|曾祖母|曾祖妣|曾大母|曾王母|曾祖姑|大王母|從祖母|繼祖母|叔祖母...   妣    葛  
1687                    ([始先早前首再次又後繼]?娶|取|配)(.{1,10}?)氏   娶    陳  
2166                    ([始先早前首再次又後繼]?娶|取|配)(.{1,10}?)氏   娶    劉  
4295  (外叔祖母|外伯祖母|曾祖母|曾祖妣|曾大母|曾王母|曾祖姑|大王母|從祖母|繼祖母|叔祖母...   母    陳  
1993                    ([始先早前首再次又後繼]?娶|取|配)(.{1,10}?)氏   娶    熊  
2678                    ([始先早前首再次又後繼]?娶|取|配)(.{1,10}?)氏   娶    李  
5831  (外叔祖母|外伯祖母|曾祖母|曾祖妣|曾大母|曾王母|曾祖姑|大王母|從祖母|繼祖母|叔祖母...   妣    趙  
73                                       (妻)(.{0,10}?)氏   妻    晁  
2708                    ([始先早前首再次又後繼]?娶|取|配)(.{1,10}?)氏  繼娶    蘇  
4846  (外叔祖母|外伯祖母|曾祖母|曾祖妣|曾大母|曾王母|曾祖姑|大王母|從祖母|繼祖母|叔祖母...   母    王

In [37]:
for index in tqdm(df_kin_extracted.index):
    sent_comp_index = df_kin_extracted.loc[index, 'sent_comp_index']
    df_temp=df_sent_comp.loc[sent_comp_index][['sent', 'sent_comp', 'nz','ns','no_noc','vno']]
    df_kin_extracted.loc[index, 'sent']=df_temp['sent']
    df_kin_extracted.loc[index, 'sent_comp']=df_temp['sent_comp']
    df_kin_extracted.loc[index, 'nz']=df_temp['nz']
    df_kin_extracted.loc[index, 'ns']=df_temp['ns']
    df_kin_extracted.loc[index, 'no_noc']=df_temp['no_noc']
    df_kin_extracted.loc[index, 'vno']=df_temp['vno']

100%|██████████| 5844/5844 [00:40<00:00, 144.82it/s]


In [38]:
len(df_kin_extracted[df_kin_extracted['name']!=''])

5844

In [39]:
df_kin_extracted[df_kin_extracted['name']!='']\
.drop(['sent_comp_index'],axis=1)\
.sort_values(by='content_id')\
.to_excel('../output/df_qsw_kin_extracted_female.xlsx', encoding='utf8')

### Son/Daughter kinship extraction.

#### TODO:
need to be splited.
- [^第][一二三四五六七八九十][男|子][^女孫].{0,150} 1566  有效比例應該有一半左右。需加入
- [^子孫孙]男(.{0,3}?)人.{0,50} 112 需加入
- (子|孫|曾孫|曾孫男|男|孫男|子男|孙男|孙)[一二三四五六七八九十]/  311 需加入

No need to be splited - results messy.
- (其兄|其弟|長兄|次兄|長弟|次弟|季兄|幼弟|同母弟|仲父|堂兄|從兄|從弟|堂弟|再從兄|從堂兄|再從弟|再從兄弟|再從昆弟|族兄|族弟|姊夫|元昆|亞弟|季弟|季兄|仲兄|仲弟|叔弟|妻兄|妻弟|嗣子|長子|次子|仲子|幼子|少子|庶子|伯子|仲子|季子|長婿|次婿|侄|侄子|從子|堂侄|再從子|再從姪|侄女婿|甥|孫|孫男|男孫|曾孫|曾孫男|曾姪孫|曾孫婿|外曾孫|玄孫|四世孫|來孫|第二子|第三子|第四子|第五子|第六子|第七子|第八子|第十子|第十三子|婿|妹婿|再從孫|孫婿|孫女婿|外孫|外孫男|姪孫|嫡孫|長孫|從孫|外甥|妻甥|甥男)[\w|\/]+.{0,2} 101新增
- (其兄|其弟|長兄|次兄|長弟|次弟|季兄|幼弟|同母弟|仲父|堂兄|從兄|從弟|堂弟|再從兄|從堂兄|再從弟|再從兄弟|再從昆弟|族兄|族弟|姊夫|元昆|亞弟|季弟|季兄|仲兄|仲弟|叔弟|妻兄|妻弟|嗣子|長子|次子|仲子|幼子|少子|庶子|伯子|仲子|季子|長婿|次婿|侄|侄子|從子|堂侄|再從子|再從姪|侄女婿|甥|孫|孫男|男孫|曾孫|曾孫男|曾姪孫|曾孫婿|外曾孫|玄孫|四世孫|來孫|第二子|第三子|第四子|第五子|第六子|第七子|第八子|第十子|第十三子|婿|妹婿|再從孫|孫婿|孫女婿|外孫|外孫男|姪孫|嫡孫|長孫|從孫|外甥|妻甥|甥男)/no_noc/.{0,2} 238新增

In [3]:
re_rule_list=[r'(曾孫男|孫男|子男|孙男|曾孫|子|孫|孙|男|女|婿)曰([^/]+?)\/',
              r'(子男)(.{0,3}?)人(\/.+)',
              r'([^女]*曾孫|孫)([^女]{0,3}?)人(\/.+)',
              r'([^女]*孫子|子)([^男女]{0,3}?)人(\/.+)',
             ]

In [4]:
kin_list=[]
df_kin_extracted=pd.DataFrame()
list_temp=[]
for re_rule in re_rule_list:
    for index in tqdm(df_sent_comp.index):
        string=df_sent_comp.loc[index, 'sent_comp']
        content_id=df_sent_comp.loc[index, 'content_id']
        content=df_sent_comp.loc[index, 'content']
        author=df_sent_comp.loc[index, 'author']
        subj=df_sent_comp.loc[index, 'subject']
        kin_list_temp=re.findall(re_rule, string)
        if kin_list_temp!=[]:
            for kin_nm in kin_list_temp:
                if len(kin_nm)==2:
                    kin_nm=(kin_nm[0], '', kin_nm[1])
                df_kin_extracted=pd.concat([pd.DataFrame([list(kin_nm)+[re_rule, string, content_id, content, author, subj]]), 
                                            df_kin_extracted], ignore_index=True)
                list_temp.append(kin_nm)

100%|██████████| 57733/57733 [00:11<00:00, 4907.17it/s]


In [5]:
df_kin_extracted.sample(3)

0  1                                                  2  \
3041  男                                                  珪〔一〕   
702   子  二  /wm/庶/wsep//no_noc//wsep/改差知/ns//wsep/廡/wsep//...   
678   子  九  /wm/鐊/wsep//no_noc//wsep//ns/績谿/no_noc//wsep/錫...   

                                             3  \
3041  (曾孫男|孫男|子男|孙男|曾孫|子|孫|孙|男|女|婿)曰([^/]+?)\/   
702            ([^女]*孫子|子)([^男女]{0,3}?)人(\/.+)   
678            ([^女]*孫子|子)([^男女]{0,3}?)人(\/.+)   

                                                      4        5  \
3041                                    男曰珪〔一〕/wsep/女未名  5801812   
702   子二人/wm/庶/wsep//no_noc//wsep/改差知/ns//wsep/廡/wse...  5888793   
678   子九人/wm/鐊/wsep//no_noc//wsep//ns/績谿/no_noc//wse...  5713244   

                                                      6    7    8  
3041  嘉定丁丑十一月丙申，廣西經略安撫司幹辦公事、同措置買馬范君克信卒於官，得年六十有九。明年七月...   劉宰  范克信  
702   公諱杲，字子昕。曾大父圯，提舉江西常平。大父鐸，知萬載縣，贈大中夫。父穎〔一〕，刑部郎中，贈...  劉克莊   杜杲  
678   亡妣姓汪氏，諱慧通，字正柔，明之鄞人。曾祖元吉不仕，祖洙，明州助教，累贈正奉大夫。父思温，左...   樓鑰  NaN

In [6]:
df_kin_extracted=df_kin_extracted.rename(columns={0:'kin', 1:'number', 2:'name_string', 
                                                  3:'re_rule', 4:'sent_comp', 5:'content_id',
                                                  6:'content', 7:'author', 8:'subj'})

In [7]:
del_dict=['皆其婿也','季曰','伯曰','仲曰','叔曰','次即','及第','中第',
          '中舉','今以','今爲','江淮','兩浙','某官','税務','支鹽',
          '士族','士人','許嫁','許適','許歸','未冠','未仕','未官',
          '未命','先歿','先亡','先公','早夭','早亡','早世','尚幼',
          '未名','前卒','先卒','蚤卒','俱有','早卒','早喪','一人',
          '二人','一尚','二尚','三尚','三女','三人','三曰','四人',
          '四先','四女','未仕','未銓','左銓','司户','户部','長',
          '次','幼','曰','即','也','貢','等','第','登','科','今',
          '授','事','都','轄','新','知','舊','監','倉','庫','起',
          '終','故','前','後','左','右','州','軍','路','郡','縣',
          '府','寺','尉','官','俱','業','業','習','適','嫁','歸',
          '許','皆','並','并','餘','俱','竝','未','夭','卒','幼',
          '尚','早','喪','一','二','三','户',]

In [8]:
def clean_name_string(name_string):
    global del_dict
    name_string_list=[s for s in re.sub(r'[a-z_]', '', name_string).split('/') if s!='']
    name_string_list_refined=[]
    for name_string in name_string_list:
        for del_string in del_dict:
            name_string=name_string.replace(del_string, '')
        if name_string!='' and len(name_string)<=2:
            name_string_list_refined.append(name_string)
    return name_string_list_refined

In [9]:
df_kin_extracted['name_string']=[clean_name_string(s) for s in df_kin_extracted['name_string']]
df_kin_extracted.sample(3)

kin number              name_string                          re_rule  \
1179  曾孫      六                     [撙之]   ([^女]*曾孫|孫)([^女]{0,3}?)人(\/.+)   
482    子      二                 [慶臣, 清臣]  ([^女]*孫子|子)([^男女]{0,3}?)人(\/.+)   
2489  子男      四  [炳文, 已, 夢得, 思義, 發解, 夢符]             (子男)(.{0,3}?)人(\/.+)   

                                              sent_comp  content_id  \
1179               曾孫六人/wsep/曰撙之/wsep//no_noc//wsep/餘尚幼     5459896   
482                                 子二人/wsep/慶臣/wsep/清臣     5770429   
2489  子男四人/wm/長炳文/wsep/已卒/wsep/次夢得/wsep/以後伯氏/wsep/凡再...     5834653   

                                                content author subj  
1179  紹興八年春，右朝議大夫、廣南東路轉運判官高君公軒奉母夫人北歸，寓居吉州之具慶坊[sub]〔一...    劉才邵  NaN  
482   余爲令臨川，愛其山川風俗之媺，名儒鉅公彬彬輩出，故家遺俗皆知尚氣節、畏清議，信厚而質直，其天...     黄榦  黄思永  
2489  孫氏之先居青社。七世祖鐐，咸平中主江陵石首簿，卒官，葬紀山皇華原，遂爲江陵人〔一〕。至四世羲...    魏了翁   孫堪

In [10]:
df_kin_extracted.to_excel('../output/df_kin_extracted_son_dau_etc.xlsx', encoding='utf8')

### Son in law.
- (孫女|女|曾孫女)[許適嫁歸].{1,200}  464 如何表达呢
- (一女|二女|三女|四女|五女|六女|七女|八女|九女|十女|十一女|十六女).{0,50}?[許適嫁歸]  913
- (曾孫女|孫女|女|女孫|妹).{0,2}人.{0,50}?[許適嫁歸] 1368
- (孫女|女|曾孫女)[一二三四五六七八九十]/   198
- (姑|同母姊|姊|妹|堂姊|堂妹|再從妹|再從姊|表妹|表姊|仲女|長女|次女|幼女|季女|甥女|表甥|表甥女|從孫女|孫婦|外甥女|曾孫女|外孫女|姪孫女|甥女)[^一二三四五六七八九十].{1,50}?[適嫁歸] 207

In [7]:
re_rule_list=[r'(孫女|女|曾孫女)[許適嫁歸](.{1,200})',
              r'(一女|二女|三女|四女|五女|六女|七女|八女|九女|十女|十一女|十六女).{0,50}?[許適嫁歸](.+?)',
              r'(曾孫女|孫女|女|女孫|妹).{0,2}人.{0,50}?[許適嫁歸](.+?)',
              r'(孫女|女|曾孫女)[一二三四五六七八九十](.+?)',
              r'(姑|同母姊|姊|妹|堂姊|堂妹|再從妹|再從姊|表妹|表姊|仲女|長女|次女|幼女|季女|甥女|表甥|表甥女|從孫女|孫婦|外甥女|曾孫女|外孫女|姪孫女|甥女)[^一二三四五六七八九十].{1,50}?[適嫁歸](.+?)',
             ]

In [8]:
kin_list=[]
df_kin_extracted=pd.DataFrame()
list_temp=[]
for re_rule in re_rule_list:
    for index in tqdm(df_sent_comp.index):
        string=df_sent_comp.loc[index, 'sent_comp']
        content_id=df_sent_comp.loc[index, 'content_id']
        content=df_sent_comp.loc[index, 'content']
        author=df_sent_comp.loc[index, 'author']
        subj=df_sent_comp.loc[index, 'subject']
        kin_list_temp=re.findall(re_rule, string)
        if kin_list_temp!=[]:
            for kin_nm in kin_list_temp:
                df_kin_extracted=pd.concat([pd.DataFrame([list(kin_nm)+[re_rule, string, content_id, content, author, subj]]), 
                                            df_kin_extracted], ignore_index=True)
                list_temp.append(kin_nm)

100%|██████████| 57733/57733 [00:08<00:00, 6560.21it/s]


In [11]:
df_kin_extracted.sample(3)

0            1                                                  2  \
4896  三女            宗  (一女|二女|三女|四女|五女|六女|七女|八女|九女|十女|十一女|十六女).{0,50}...   
111    姑            于  (姑|同母姊|姊|妹|堂姊|堂妹|再從妹|再從姊|表妹|表姊|仲女|長女|次女|幼女|季女|...   
5484   女  /no_noc/周景南                         (孫女|女|曾孫女)[許適嫁歸](.{1,200})   

                                                      3        4  \
4896                            三女/wsep/長嫁宗室崇譓/wsep/二尚幼  5764623   
111   今吾外姑之葬/wsep/雖微諸孤之請/wsep/吾敢無銘乎?外姑踰笄/wsep/歸于外舅/w...  5731174   
5484  娶陳氏/wsep/生文與方及一女而卒/wsep/及見方/vno//no_noc//wsep/...  5784081   

                                                      5    6    7  
4896  嗚呼，學之難也!質與性不合，年與人不並，幼而教之，時至而不得成其材，十且八九也，況不幸失教者...   葉適  丁希亮  
111   余少時見墓銘日多，往往緣稱美之義，不復顧其實，侈言溢辭，使人無取信。竊念之曰：‘苟如是，不如...  陸九淵  NaN  
5484  梅坡先生之喪，當有銘以納諸壙。其子方會稡行實，乃以屬諸某，謹論次其事。先生諱蠡，字師范，避大...  曹彦約  彭蠡、

In [12]:
df_kin_extracted=df_kin_extracted.rename(columns={0:'kin', 1:'name_string', 
                                                  2:'re_rule', 3:'sent_comp', 4:'content_id',
                                                  5:'content', 6:'author', 7:'subj'})

In [13]:
df_kin_extracted.sample(3)

kin                name_string  \
5161   女  /vno//no_noc/劉庠/wsep/次女在室   
3329   女                          /   
2059   女                          人   

                                        re_rule  \
5161                 (孫女|女|曾孫女)[許適嫁歸](.{1,200})   
3329  (曾孫女|孫女|女|女孫|妹).{0,2}人.{0,50}?[許適嫁歸](.+?)   
2059                (孫女|女|曾孫女)[一二三四五六七八九十](.+?)   

                                              sent_comp  content_id  \
5161                       長女適/vno//no_noc/劉庠/wsep/次女在室     5017311   
3329  女子七人/wm/適/no_noc/焦莊叔/wsep/修/ns/趙叔蘧/wsep//no_no...     5296548   
2059  女二人/wsep/長適/no_noc/費伯午/wsep/次適/no_noc//wsep/監/...     5834753   

                                                content author subj  
5161  寶元二年，歲次己卯，四月，前參知政事、户部侍郎蔡公薨，天子悼之，卿大夫憂之，國人傷之。上命三...    范仲淹   蔡齊  
3329  　　公諱公述，字傳之，丞相文元公之孫，尚書駕部員外郎、贈右光禄大夫諱田之子也。文元公以經術勳...     王寔  NaN  
2059  慶元末，予試吏西川幕府，與郭君幼才爲僚。暨守眉，幼才爲青神宰，戢姦弭訟，崇學厲俗，有治理聲，...    魏了翁  郭叔誼

In [14]:
del_dict=['皆其婿也','季曰','伯曰','仲曰','叔曰','次即','及第','中第',
          '中舉','今以','今爲','江淮','兩浙','某官','税務','支鹽',
          '士族','士人','許嫁','許適','許歸','未冠','未仕','未官',
          '未命','先歿','先亡','先公','早夭','早亡','早世','尚幼',
          '未名','前卒','先卒','蚤卒','俱有','早卒','早喪','一人',
          '二人','一尚','二尚','三尚','三女','三人','三曰','四人',
          '四先','四女','未仕','未銓','左銓','司户','户部','長',
          '次','幼','曰','即','也','貢','等','第','登','科','今',
          '授','事','都','轄','新','知','舊','監','倉','庫','起',
          '終','故','前','後','左','右','州','軍','路','郡','縣',
          '府','寺','尉','官','俱','業','業','習','適','嫁','歸',
          '許','皆','並','并','餘','俱','竝','未','夭','卒','幼',
          '尚','早','喪','一','二','三','户',]

In [15]:
def clean_name_string(name_string):
    global del_dict
    name_string_list=[s for s in re.sub(r'[a-z_]', '', name_string).split('/') if s!='']
    name_string_list_refined=[]
    for name_string in name_string_list:
        for del_string in del_dict:
            name_string=name_string.replace(del_string, '')
        if name_string!='' and len(name_string)<=2:
            name_string_list_refined.append(name_string)
    return name_string_list_refined

In [16]:
df_kin_extracted['name_string']=[clean_name_string(s) for s in df_kin_extracted['name_string']]
df_kin_extracted.sample(3)

In [19]:
df_kin_extracted.to_excel('../output/df_kin_extracted_son_in_law.xlsx', encoding='utf8')